Add enum validation to `DomainBlock#severity`

hendrixfan opened the pull request:
Closes https://github.com/mastodon/mastodon/issues/29126

[Rails 7.1](https://github.com/rails/rails/pull/49100) introduced a validation option for enums. By enabling validation for DomainBlock#severity a `ActiveRecord::RecordInvalid` is thrown on `create!` if an invalid value is assigned. Therefore, the API returns 422.

hendrixfan committed:
Add severity validation

mjankowski reviewed (approved):
Looks good overall. A few ideas:

- Do we need to care at all about the edge case where there could be records which were previously wrong-but-valid (ie, had DB values outsize of `0,1,2`) and will now also be invalid? These could only have been set via direct sql, as using the setters would have previously errored if they were given symbols that were not valid enum keys, so maybe we can just not care. Mitigation here would be like a migration which changed any value higher than 2 to be a 2, or similar.
- I note the signature/style change here necessary to get the feature to work -- might make sense to do a once over on some other enum declarations and update to use this new style (I think the previous may be deprecated in future rails)
- Might also be useful to do a once-over on other enum values which have similar API access or other access where we could improve an error condition with the same sort of change.

mjankowski commented on spec/requests/api/v1/admin/domain_blocks_spec.rb:
```diff
@@ -168,6 +168,16 @@
+      it 'returns http unprocessable entity' do
+        subject
+
+        expect(response).to have_http_status(422)
```
Could we also add an assertion about the error json returned in the response?

hendrixfan replied:
Yeah good idea. I will add one

hendrixfan replied:
Check. I've added an assertion about the returned JSON.

hendrixfan commented:
The old way of defining will be deprecated very soon. A PR has already been merged https://github.com/rails/rails/pull/50987 and, as @mjankowski correctly stated, `validate: true` is only supported with positional arguments.

hendrixfan commented:
I think it would not hurt anyone to do a migration here. As @mjankowski stated, `DomainBlock#severity` should stay within a valid range. The impact of using non-valid ENUM values changes with PR and should be addressed.

hendrixfan committed:
fix invalid domain block severities

This commit fixes any invalid domain block severities by setting them to
2 if they are bigger than 2 or 0 if they are smaller than 0.
This ensures that the domain block severities are always within the valid
range of 0 to 2.

mjankowski commented:
The failing checks here don't seem relevant to PR ... want to rebase and let CI run against latest config?

hendrixfan committed:
Update spec to match upstream response expectations

mjankowski commented:
Hmm. CI failure here is intermittent issue unrelated to changes.

hendrixfan commented:
@mjankowski Thanks for having look at this PR again! I kinda forgot about it.

mjankowski commented:
Resolves https://github.com/mastodon/mastodon/issues/21775 (I think) as well

mjankowski reviewed (approved):
This looks fine to me, and has been added to my [top ten list](https://github.com/mjankowski/mastodon/wiki/TOPTEN).

I'm not sure how to get someone with merge permissions to review.

ClearlyClaire commented:
Hi! Sorry we haven't reviewed it earlier. The current behavior is that an invalid argument raises `ArgumentError`, so invalid severities cannot end up in the database through the application code.

Presence of other values can only be caused by manual operation from the administrator, or custom code changes.

Therefore, I think the migration is unnecessary, and more trouble than is helpful.

hendrixfan committed:
Delete db/migrate/20240220082519_fix_invalid_domain_block_severities.rb

ClearlyClaire reviewed (approved):
Looks ok to me, only thing to double-check is that we are never bypassing validation when saving domain blocks.
